File: files-app/src/main/java/l/files/fs/local/LocalPathEntry.java
package l.files.fs.local;

import java.io.File;
import java.io.IOException;

import auto.parcel.AutoParcel;
import l.files.fs.PathEntry;

@AutoParcel
abstract class LocalPathEntry implements PathEntry {

    LocalPathEntry() {
    }

    @Override
    public abstract LocalResource getResource();

    public abstract long getInode();

    public abstract boolean isDirectory();

    @Override
    public LocalPath getPath() {
        return getResource().getPath();
    }

    public static LocalPathEntry create(LocalResource resource, long inode, boolean directory) {
        return new AutoParcel_LocalPathEntry(resource, inode, directory);
    }

    public static LocalPathEntry stat(File file) throws IOException {
        return read(LocalResource.create(file));
    }

    public static LocalPathEntry read(LocalResource resource) throws IOException {
        LocalResourceStatus status = LocalResourceStatus.stat(resource.getFile(), false);
        return create(resource, status.getInode(), status.isDirectory());
    }

    // observerable

}

File: files-app/src/main/kotlin/fs/local/LocalPathEntry.kt
package l.files.fs.local

import l.files.fs.Path
import l.files.fs.PathEntry
import java.io.IOException
import java.io.File

private data class LocalPathEntry(
        private val _path: LocalPath,
        val ino: Long,
        val isDirectory: Boolean) : PathEntry {

    override fun getPath() = _path

    override fun getResource() = getPath().getResource()

    public companion object {

        throws(javaClass<IOException>())
        fun stat(file: File): LocalPathEntry {
            return read(LocalPath.of(file))
        }

        throws(javaClass<IOException>())
        fun read(path: LocalPath): LocalPathEntry {
            val status = LocalResourceStatus.stat(path, false)
            return LocalPathEntry(path, status.inode, status.isDirectory)
        }

    }

}
